### PR TITLE
Feature/calculate energy needed

### PR DESCRIPTION
**Pull request by hugocowan**
Made a system for calculating exactly how much stamina will get used for the next run as you are running back to the bank. It ensures you get the most out of your stamina potions, making the process more efficient and saving a bit of money too.

It is linked to the Low energy threshold config setting - it will not allow your stamina to be less than the threshold.

Things it does:

- Calculates run energy required ahead of time for each part of the run to and from the BF, including the 9 ticks to take ore to the BF, the 5 ticks you'll have running with an empty inventory (9 if it's a coal run), and finally the 4 ticks at the end when you will be holding bars.

- Dynamically accounts for what the player is wearing, what they will need to carry for the next run to the BF and back, and how many ticks they'll be idle during the run.

- Proportionately limits effectiveness of stamina potion when its timer will run out before you make it back to the bank.

The end result is pretty awesome. Even with the threshold set to 0, you won't run out of run energy, and stamina potion timers are used up as much as possible before another is highlighted.

### CODE REVIEW

File: src/main/java/com/toofifty/easyblastfurnace/state/InventoryState.java [hunk 1]
```diff
@@ -68,13 +70,20 @@ public boolean has(int ...itemIds) {
...
         int freeSlots = 28;
+        int[] barsAndOres = new int[]{
+            ItemID.IRON_BAR, ItemID.MITHRIL_BAR, ItemID.ADAMANTITE_BAR, ItemID.RUNITE_BAR, ItemID.GOLD_BAR, ItemID.STEEL_BAR,
+            ItemID.IRON_ORE, ItemID.MITHRIL_ORE, ItemID.ADAMANTITE_ORE, ItemID.RUNITE_ORE, ItemID.GOLD_ORE, ItemID.COAL
+        };
+
         for (Item item : inventory.getItems()) {
-            if (item.getQuantity() > 0) {
+            if (ignoreBarsAndOres && IntStream.of(barsAndOres).noneMatch(id -> id == item.getId()) && item.getQuantity() > 0) {
+                freeSlots--;
```
Comment (Toofifty) [inline src/main/java/com/toofifty/easyblastfurnace/state/InventoryState.java:85]: Is this _really_ still the amount of free slots? I feel like the new calculation should be it's own method

File: src/main/java/com/toofifty/easyblastfurnace/state/PlayerState.java [hunk 2]
```diff
@@ -30,10 +30,8 @@
     @Inject
     private EasyBlastFurnaceConfig config;
 
-    public int getRunEnergy()
-    {
-        return client.getEnergy();
-    }
+    @Inject
+    StaminaHelper staminaHelper;
```
Comment (Toofifty) [inline src/main/java/com/toofifty/easyblastfurnace/state/PlayerState.java:34]:
```suggestion
    private StaminaHelper staminaHelper;
```

File: src/main/java/com/toofifty/easyblastfurnace/utils/StaminaHelper.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import net.runelite.api.*;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.util.RSTimeUnit;
+
+import javax.inject.Inject;
+import java.time.Duration;
+import java.time.Instant;
+
+@Slf4j
+public class StaminaHelper {
```
Comment (Toofifty) [inline src/main/java/com/toofifty/easyblastfurnace/utils/StaminaHelper.java:15]: This is so over-engineered. I love it

File: src/main/java/com/toofifty/easyblastfurnace/state/InventoryState.java (hunk 1 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/state/InventoryState.java]: Fair, I've made a separate method `getFreeSlotsIncludingOresAndBars` to address this.

File: src/main/java/com/toofifty/easyblastfurnace/state/PlayerState.java (hunk 2 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/state/PlayerState.java]: Oops

File: src/main/java/com/toofifty/easyblastfurnace/utils/StaminaHelper.java (hunk 3 above)
Comment (hugocowan) [reply src/main/java/com/toofifty/easyblastfurnace/utils/StaminaHelper.java:15]: Thanks lol

**Comment by hugocowan** [comment]
@Toofifty This is good to go from my perspective :) Lemme know if you need anything else changing.

**Pull request merged**